### PR TITLE
DDF-1852: Added data usage tracking on resource requests

### PR DESCRIPTION
**Pull request by jlcsmith**
- Added a PreResourcePlugin to keep track of data usage per user
- Added a new solr core for user attributes
- Added persistent store interface for attributes

@jrnorth

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/codice/ddf/pull/545" target="_blank">ReviewNinja

File: catalog/plugin/catalog-plugin-resource-usage/src/test/java/ddf/catalog/plugin/resource/usage/TestResourceUsagePlugin.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+        ResourceRequest origRequest = getMockResourceRequest(RESOURCE_SIZE, TEST_USER);
+        ResourceRequest request = plugin.process(origRequest);
+
+        assertThat(request, notNullValue());
+        assertThat(request, is(equalTo(origRequest)));
+
+        verify(attributeStore).updateUserDataUsage(usernameArg.capture(), dataSizeArg.capture());
+
+        assertThat(usernameArg.getValue(), is(equalTo(TEST_USER)));
```
Comment (jrnorth) [inline catalog/plugin/catalog-plugin-resource-usage/src/test/java/ddf/catalog/plugin/resource/usage/TestResourceUsagePlugin.java]: `is(equalTo())` can be simplified to [`is()`](http://hamcrest.org/JavaHamcrest/javadoc/1.3/org/hamcrest/core/Is.html#is(T)).

File: catalog/plugin/catalog-plugin-resource-usage/src/test/java/ddf/catalog/plugin/resource/usage/TestResourceUsagePlugin.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                    @Override
+                    public String toString() {
+                        return expectedUsername;
+                    }
+                }, realm.getName());
+
+        subject = new MockSubject(manager, principalCollection);
+
+        ResourceRequest resourceRequest = mock(ResourceRequest.class);
+        Map<String, Serializable> requestProperties = new HashMap<String, Serializable>();
```
Comment (jrnorth) [inline catalog/plugin/catalog-plugin-resource-usage/src/test/java/ddf/catalog/plugin/resource/usage/TestResourceUsagePlugin.java]: The type parameters for the `HashMap` aren't necessary.

File: platform/persistence/platform-persistence-core-attributes-impl/src/main/java/org/codice/ddf/persistence/attributes/internal/AttributesStoreImpl.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        if (StringUtils.isEmpty(username)) {
+            throw new PersistenceException(EMPTY_USERNAME_ERROR);
+        }
+        if (dataUsage >= 0) {
+            try {
+                readWriteLock.writeLock()
+                        .lock();
+
+                // add to usage and store
+                PersistentItem item = new PersistentItem();
```
Comment (jrnorth) [inline platform/persistence/platform-persistence-core-attributes-impl/src/main/java/org/codice/ddf/persistence/attributes/internal/AttributesStoreImpl.java]: This code is identical to the code in `updateUserDataUsage()`.

Comment (jlcsmith) [inline platform/persistence/platform-persistence-core-attributes-impl/src/main/java/org/codice/ddf/persistence/attributes/internal/AttributesStoreImpl.java]: !fixed

File: catalog/plugin/catalog-plugin-resource-usage/src/test/java/ddf/catalog/plugin/resource/usage/TestResourceUsagePlugin.java (hunk 2 above)
Comment (jlcsmith) [inline catalog/plugin/catalog-plugin-resource-usage/src/test/java/ddf/catalog/plugin/resource/usage/TestResourceUsagePlugin.java]: !fixed

File: catalog/plugin/catalog-plugin-resource-usage/src/test/java/ddf/catalog/plugin/resource/usage/TestResourceUsagePlugin.java (hunk 1 above)
Comment (jlcsmith) [inline catalog/plugin/catalog-plugin-resource-usage/src/test/java/ddf/catalog/plugin/resource/usage/TestResourceUsagePlugin.java]: !fixed

File: platform/persistence/platform-persistence-core-attributes-impl/src/test/java/org/codice/ddf/persistence/attributes/internal/TestAttributesStoreImpl.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        attributes.put(DATA_USAGE_LONG, 100L);
+        attributesList.add(attributes);
+        when(persistentStore.get(any(String.class), any(String.class))).thenReturn(attributesList);
+
+        attributesStore.updateUserDataUsage(USER, 500L);
+
+        verify(persistentStore).get(keyArg1.capture(), cqlArg.capture());
+        verify(persistentStore).add(keyArg2.capture(), itemArg.capture());
+
+        assertThat(keyArg1.getValue(), is(equalTo(PersistentStore.USER_ATTRIBUTE_TYPE)));
```
Comment (jrnorth) [inline platform/persistence/platform-persistence-core-attributes-impl/src/test/java/org/codice/ddf/persistence/attributes/internal/TestAttributesStoreImpl.java]: These could also be `is()`.

**Comment by rzwiefel** [comment]
:+1:

**Pull request merged**